lib/fog/orchestration/util/recursive_hot_file_loader.rb

```diff
@@ -0,0 +1,227 @@
+                         !value.end_with?('.yaml', '.template')
+
+          false
+        end
+
+        def recurse_if(value)
+          # Should I recurse into this template branch?
+          value.kind_of?(Hash) || value.kind_of?(Array)
+        end
+
+        def url_join(prefix, suffix)
+          # return string
+          if prefix
+            suffix = URI.join(prefix, suffix)
+            fixup_uri(suffix)
+            suffix = suffix.to_s
+          end
+          suffix
+        end
+
+        def base_url_for_url(url)
+          # Returns the string baseurl of the given url.
+          parsed = URI(url)
+          # Ruby URI is not round-trip safe when schema == file
+          fixup_uri(parsed)
+          parsed_dir = File.dirname(parsed.path)
+          URI.join(parsed, parsed_dir).to_s
+        end
+
+        def normalise_file_path_to_url(path)
+          # Nothing to do on URIs
+          return path if URI(path).scheme
+
+          path = File.absolute_path(path)
+          url_join('file:', path)
+        end
+
+        def get_content(uri_or_filename)
+          Fog::Logger.warning("Opening #{uri_or_filename}")
+          uri_or_filename = fixup_uri(uri_or_filename)
+          uri_or_filename = uri_or_filename.to_s if uri_or_filename.kind_of?(URI)
+          # throw exceptions enables stack creation to fail
+          #   with a suitable error.
+          #
+          # XXX Implement a retry here?
+          content = ''
+          # does ruby close the socket/filedescriptor at exit?
+          # XXX Limit download file size
+          # XXX Protect from vanilla open-uri attacks
+          uri_or_filename = uri_or_filename[7..-1] if uri_or_filename.start_with?("file:///")
+          open(uri_or_filename) { |f| content = f.read }
+          content == "Error" ? nil : content
+
+          content
+        end
+
+        def get_template_contents(template_file, _template_url = nil, _template_object = nil, _existing = false)
+          # Same code for both template_file and template_url.
+          Fog::Logger.warning("get_template_contents #{template_file}")
+
+          local_base_url = base_url_for_url(normalise_file_path_to_url(Dir.pwd + "/TEMPLATE"))
+          if template_file.kind_of?(String)
+            if is_template(template_file)
+              tpl = template_file
+              template_base_url = local_base_url
+            else
+              template_file = normalise_file_path_to_url(template_file)
+              template_base_url = base_url_for_url(template_file)
+              # TODO: normalize template_file
+              tpl = get_content(template_file)
+              @visited[template_file] = true
+              Fog::Logger.warning("Template visited: #{@visited}")
+            end
+            template = YAML.load(tpl)
+          elsif template_file.kind_of?(Hash)
+            # Normalize yaml.
+            Fog::Logger.warning("Reingest")
+            template = YAML.safe_load(YAML.dump(template_file))
```

Comment:
Why do you need to normalize yaml? Here `template_file` is already a hash, isn't it?